File: src/2015/AdventOfCode2015/Day02.cs
using AdventOfCodeLib;

namespace AdventOfCode2015;

[AocPuzzle(2015, 2, "I Was Told There Would Be No Math")]
public sealed class Day02 : IAocDay<int>
{
    public static int Part1(AocInput input) => SizeUpPresents(input.Lines)
        .Sum(x => OrderWrappingPaper(x.Length, x.Width, x.Height));

    public static int Part2(AocInput input) => SizeUpPresents(input.Lines)
        .Sum(x => OrderRibbon(x.Length, x.Width, x.Height));

    private static IEnumerable<Present> SizeUpPresents(IEnumerable<string> input) =>
        input.Select(x => x.Split('x').Select(int.Parse).ToArray())
            .Select(x => new Present(x[0], x[1], x[2]));

    private static int OrderWrappingPaper(int length, int width, int height) =>
        CalculateSurface(length, width, height) + CalculateAreaOfSmallestSide(length, width, height);

    private static int OrderRibbon(int length, int width, int height) =>
        CalculateBow(length, width, height) + CalculatePerimeterOfSmallestSide(length, width, height);

    private static int CalculateSurface(int length, int width, int height) =>
        2 * length * width + 2 * width * height + 2 * height * length;

    private static int CalculateAreaOfSmallestSide(int length, int width, int height) =>
        Math.Min(Math.Min(length * width, width * height), height * length);

    private static int CalculateBow(int length, int width, int height) =>
        length * width * height;

    private static int CalculatePerimeterOfSmallestSide(int length, int width, int height) =>
        Math.Min(Math.Min(2 * length + 2 * width, 2 * width + 2 * height), 2 * height + 2 * length);

    private readonly record struct Present(int Length, int Width, int Height);
}

File: python/y2015/day02.py
from aoc.core import solve, AocInput
from dataclasses import astuple, dataclass


@dataclass(frozen=True)
class Present:
    length: int
    width: int
    height: int

    def __iter__(self):
        return iter(astuple(self))


def makePresent(values: list[str]) -> Present:
    return Present(int(values[0]), int(values[1]), int(values[2]))


def sizeUpPresents(lines: list[str]) -> list[Present]:
    return [makePresent(l.split("x")) for l in lines]


def orderWrapping(present: Present) -> int:

    def calculateSurface(l: int, w: int, h: int) -> int:
        return 2 * l * w + 2 * w * h + 2 * h * l

    def calculateAreaOfSmallestSide(l: int, w: int, h: int) -> int:
        return min(min(l * w, w * h), h * l)

    length, width, height = present
    return calculateSurface(length, width, height) + calculateAreaOfSmallestSide(length, width, height)


def orderRibbon(present: Present) -> int:

    def calculateBow(l: int, w: int, h: int) -> int:
        return l * w * h

    def calculatePerimeterOfSmallestSide(l: int, w: int, h: int) -> int:
        return min(min(2 * l + 2 * w, 2 * w + 2 * h), 2 * h + 2 * l)

    length, width, height = present
    return calculateBow(length, width, height) + calculatePerimeterOfSmallestSide(length, width, height)


def part1(input: AocInput) -> int:
    return sum(map(orderWrapping, sizeUpPresents(input.lines)))


def part2(input: AocInput) -> int:
    return sum(map(orderRibbon, sizeUpPresents(input.lines)))


def main() -> None:
    YEAR = "y2015"
    DAY = "Day02"
    print(f"Running year {YEAR.replace('y')} day {DAY.replace('Day')}")
    solve(YEAR, DAY, part1)
    solve(YEAR, DAY, part2)


if __name__ == "__main__":
    main()
